File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys

from collections import deque
input = sys.stdin.readline


def bfs(v):
    queue = deque([v])
    visited[v] = 1
    while queue:
        x = queue.popleft()
        for i in rel[x]:
            if not visited[i]:
                queue.append(i)
                visited[i] = visited[x] + 1
            
            

n = int(input())
a,b = map(int,input().split())

m = int(input())

rel = [[] for _ in range(n+1)]
visited = [0] * (n+1)
for _ in range(m):
    x, y = map(int,input().split())
    rel[x].append(y)
    rel[y].append(x)

bfs(min(a,b))
print(visited[max(a,b)]-visited[min(a,b)])

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.io.IOException;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;


public class Main {
    static int n,a,b,m,x,y;
    static StringTokenizer st;
    static List<Integer>[] rel;
    static int[] visited;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        a = Integer.parseInt(st.nextToken());
        b = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(br.readLine());
        
        rel = new ArrayList[n+1];
        for(int i =1; i<n+1; i ++){
            rel[i] = new ArrayList<>();
        }
        visited = new int[n+1];
        for(int i =0; i< m; i ++){
            st = new StringTokenizer(br.readLine());
            x = Integer.parseInt(st.nextToken());
            y = Integer.parseInt(st.nextToken());
            rel[x].add(y);
            rel[y].add(x);
        }
        bfs(Math.min(a,b));
        System.out.println(visited[Math.max(a,b)] - visited[Math.min(a,b)]);
    }

    public static void bfs(int v){
        Queue<Integer> queue = new LinkedList<Integer>();
        queue.add(v);
        visited[v] = 1;
        while(!queue.isEmpty()){
            x = queue.poll();
            for(int i : rel[x]){
                if(visited[i] == 0){
                    queue.add(i);
                    visited[i] = visited[x] + 1;
                }
            }
        }
    }
} 
